src/sentry/static/sentry/app/components/asyncComponent.tsx

```diff
@@ -68,7 +66,7 @@ function wrapErrorHandling<T extends any[], U>(
 }
 
 export default class AsyncComponent<
-  P extends AsyncComponentProps = AsyncComponentProps,
+  P extends AsyncComponentProps & RouteProps = AsyncComponentProps,
```

Comment:
I actually don't what this syntax (`AsyncComponentProps & RouteProps = AsyncComponentProps`) represents, I just know it does what I want it to do. I tried finding documentation on this pattern but I couldn't find anything. Would appreciate if anyone could point me to the documentation or explanation what the hell this is.

Reply:
`AsyncComponentProps & RouteProps` - this is called "type intersection"; it's a way to combine/merge types together. https://www.typescriptlang.org/docs/handbook/advanced-types.html#intersection-types

------

This part `P = AsyncComponentProps` is the default type for `P` -
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults

The default parameter increases the ergonomics of `AsyncComponent`; i.e. not having to explicitly define the type for `P`.

Reply:
@dashed Thanks for the explanation. What I am most confused about is the fact the default is on the left hand side of the equality, not the right as usual for Javascript. Here's what I would expect would work:
```
AsyncComponentProps = AsyncComponentProps & RouteProps
```
As opposed to what actually works:
```
AsyncComponentProps & RouteProps = AsyncComponentProps
```

Reply:
Disclaimer: this is my explanation without running any TS code.

For the first example (where you expected it to work), 

```
P extends AsyncComponentProps = AsyncComponentProps & RouteProps

1. P extends AsyncComponentProps
2. Assume P is of type AsyncComponentProps & RouteProps (the default type)
3. (AsyncComponentProps & RouteProps) extends AsyncComponentProps is invalid
```

It's invalid because `AsyncComponentProps & RouteProps` cannot be a "subtype" of `AsyncComponentProps`; since `AsyncComponentProps & RouteProps` may contain more "stuff" than `AsyncComponentProps`.

----

For the second, `AsyncComponentProps` can be a subtype of `AsyncComponentProps & RouteProps`.

Reply:
@dashed Interesting theory. The reason it fails is because of type errors like this:

<img width="547" alt="Screen Shot 2019-12-26 at 7 39 33 PM" src="https://user-images.githubusercontent.com/8533851/71499919-948f4e80-2817-11ea-97e7-daffa61702d0.png">

There's not actually an error with the generic subtyping/extension/whatever

I think the key is that the default value is _not_ used when evaluating the type of the component, it's what's to the left of the equality (`AsyncComponentProps & RouteProp`).

Anyways, I could simplify this by making the default value the same as the base type of hte component. I was able to remove all of the `& AsyncView['props']` in our code. But for some reason, I can't remove the  `& AsyncComponent['props']` logic in our code. Go figure.